### PR TITLE
[MLIR][ONNX] Add reduceprod op

### PR DESCRIPTION
**Pull request by AmosLewis**
https://github.com/nod-ai/SHARK-Turbine/issues/353

### CODE REVIEW

**Comment by renxida** [comment]
https://gist.github.com/renxida/823ea06a550fd3c79a11008597cca3b9

I got really annoyed by having to run clang-format.

Save this as `.git/hooks/pre-commit` to automatically run clang-format at every commit

**Comment by AmosLewis** [comment]
specify axes
```
torch-mlir-opt -convert-torch-onnx-to-torch ../onnx_torch_import_20231121/node_test_reduce_prod_negative_axes_keepdims_example_model.mlir
module {
  func.func @test_reduce_prod_negative_axes_keepdims_example(%arg0: !torch.vtensor<[3,2,2],f32>, %arg1: !torch.vtensor<[1],si64>) -> !torch.vtensor<[3,1,2],f32> attributes {torch.onnx_meta.ir_version = 8 : si64, torch.onnx_meta.opset_version = 18 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
    %none = torch.constant.none
    %int0 = torch.constant.int 0
    %int3 = torch.constant.int 3
    %int1 = torch.constant.int 1
    %0 = torch.aten.Bool.int %int1 : !torch.int -> !torch.bool
    %int0_0 = torch.constant.int 0
    %1 = torch.aten.select.int %arg1, %int0, %int0_0 : !torch.vtensor<[1],si64>, !torch.int, !torch.int -> !torch.vtensor<[1],si64>
    %2 = torch.aten.item %1 : !torch.vtensor<[1],si64> -> !torch.int
    %3 = torch.aten.lt.int %2, %int0 : !torch.int, !torch.int -> !torch.bool
    %4 = torch.aten.Int.bool %3 : !torch.bool -> !torch.int
    %5 = torch.aten.mul.int %4, %int3 : !torch.int, !torch.int -> !torch.int
    %6 = torch.aten.add.int %2, %5 : !torch.int, !torch.int -> !torch.int
    %7 = torch.aten.prod.dim_int %arg0, %6, %0, %none : !torch.vtensor<[3,2,2],f32>, !torch.int, !torch.bool, !torch.none -> !torch.vtensor<[3,1,2],f32>
    return %7 : !torch.vtensor<[3,1,2],f32>
  }
}
```

default axes
```
torch-mlir-opt -convert-torch-onnx-to-torch ../onnx_torch_import_20231121/node_test_reduce_prod_default_axes_keepdims_random_model.mlir
!torch.vtensor<[1],f32>
module {
  func.func @test_reduce_prod_default_axes_keepdims_random(%arg0: !torch.vtensor<[3,2,2],f32>) -> !torch.vtensor<[1,1,1],f32> attributes {torch.onnx_meta.ir_version = 8 : si64, torch.onnx_meta.opset_version = 18 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
    %none = torch.constant.none
    %int0 = torch.constant.int 0
    %int1 = torch.constant.int 1
    %0 = torch.aten.Bool.int %int1 : !torch.int -> !torch.bool
    %int12 = torch.constant.int 12
    %1 = torch.prim.ListConstruct %int12 : (!torch.int) -> !torch.list<int>
    %2 = torch.aten.reshape %arg0, %1 : !torch.vtensor<[3,2,2],f32>, !torch.list<int> -> !torch.vtensor<[12],f32>
    %3 = torch.aten.prod.dim_int %2, %int0, %0, %none : !torch.vtensor<[12],f32>, !torch.int, !torch.bool, !torch.none -> !torch.vtensor<[1],f32>
    %int1_0 = torch.constant.int 1
    %int1_1 = torch.constant.int 1
    %int1_2 = torch.constant.int 1
    %4 = torch.prim.ListConstruct %int1_0, %int1_1, %int1_2 : (!torch.int, !torch.int, !torch.int) -> !torch.list<int>
    %5 = torch.aten.reshape %3, %4 : !torch.vtensor<[1],f32>, !torch.list<int> -> !torch.vtensor<[1,1,1],f32>
    return %5 : !torch.vtensor<[1,1,1],f32>
  }
}
```

**Comment by AmosLewis** [comment]
Rob suggests a new way to deal with multiple dim once for all:  invoke torch.prod.int multiple times, i.e. for each dimension.
Here is the python code:

```
input = torch.tensor(
  [
    [[1,2,3,4],[1,2,3,4],[1,2,3,4]],
    [[1,2,3,4],[1,2,3,4],[1,2,3,4]]
  ]
  ) # torch.Size([2, 3, 4])
print('input.shape: ', input.shape)
print('input: ', input)

prod0 = torch.prod(input,0,keepdim=True)
print('prod0.shape: ', prod0.shape) 
print('prod0: ', prod0)
# prod0.shape:  torch.Size([1, 3, 4])
# prod0:  tensor([[[ 1,  4,  9, 16],
#          [ 1,  4,  9, 16],
#          [ 1,  4,  9, 16]]])

prod1 = torch.prod(prod0,1, keepdim=True)
print('prod1.shape: ', prod1.shape)
print('prod1: ', prod1)
# prod1.shape:  torch.Size([1, 1, 4])
# prod1:  tensor([[[   1,   64,  729, 4096]]])

prod2 = torch.prod(prod1,2, keepdim=True)
print('prod1.shape: ', prod2.shape)
print('prod1: ', prod2)
# prod1.shape:  torch.Size([1, 1, 1])
# prod1:  tensor([[[191102976]]])
```

Rob also suggests for default axes, manually create a axes tensor for it.
```
llvm::SmallVector<int64_t> axisToReduce;
if (operands.size() == 1) {
  for (int i = 0; i < rank; i++) {
    axisToReduce.push_back(i);
  }
} else {
  axisToReduce.push_back(Constant(operands[1]));
}
```
